legate/core/solver.py

```diff
@@ -476,45 +541,28 @@ def cost(unknown: PartSym) -> tuple[int, bool]:
                 not store.has_key_partition(all_restrictions[unknown]),
             )
 
-        sorted_unknowns = sorted(unknowns, key=cost)
-
-        key_parts = set()
-        for unknown in sorted_unknowns:
-            if unknown in partitions:
-                continue
-            elif unknown in dependent:
-                continue
-
-            store = unknown.store
-            restrictions = all_restrictions[unknown]
-            cls = constraints.find(unknown)
-
-            partition = store.compute_key_partition(restrictions)
-            if not partition.even and len(cls) > 1:
-                partition, unknown = self.maybe_find_alternative_key_partition(
-                    partition,
-                    unknown,
-                    cls,
-                    restrictions,
-                    must_be_even,
-                )
-            key_parts.add(unknown)
+        result: dict[PartSym, PartitionBase]
+        key_parts: set[PartSym]
+        launch_shape: Optional[Shape]
+
+        can_retry = True
+        while True:
+            result, key_parts = self._solve_store_constraints(
+                partitions,
+                sorted(unknowns, key=cost),
+                dependent,
+                all_restrictions,
+                constraints,
+                must_be_even,
+            )
 
-            for to_align in cls:
-                if to_align in partitions:
+            launch_shape, done = self.compute_launch_shape(
+                result, all_outputs, unbound_ndim
+            )
+            if can_retry and not done:
```

Comment:
This is a bit of tricky logic that could use a comment.

Reply:
https://github.com/nv-legate/legate.core/pull/510/commits/595965ec41013f28447c8294b7f8811ff327c763